### DISCUSSION TITLE
problem downloading examples, version 2.1.5

### DISCUSSION THREAD

**Discussion by NickGotts** [mesa/mesa#1922] [Q&A]
Hi,
I've installed mesa version 2.1.5 in an Anaconda Python 3.9 environment, and then tried to follow the instruction from https://mesa.readthedocs.io/en/stable/index.html#using-mesa which reads:
"To launch an example model, clone the [repository](https://github.com/projectmesa/mesa) folder and invoke mesa runserver for one of the examples/ subdirectories"
I've cloned the repository, apparently successfully, but no examples/ directory appears. What am I doing wrong, or failing to do?

**Comment by EwoutH** [comment]
I think those docs are out of date and should be updated. The examples were moved to the [mesa-examples](https://github.com/projectmesa/mesa-examples) repository in #1529.

**Reply by NickGotts** [reply]
Thanks for that - it enabled me to downlaod the examples - but they still won't run! Wherever I put the mesa-examples directory, I get this error message when I try to run one of them:

File "C:\Users\Nick\anaconda3\envs\mesaenv39\lib\runpy.py", line 197, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\lib\runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\Scripts\mesa.exe\__main__.py", line 4, in <module>
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\src\mesa\mesa\main.py", line 8, in <module>
    from mesa import __version__
ImportError: cannot import name '__version__' from 'mesa' (unknown location)

**Comment by NickGotts** [comment]
On further investigation, it seems clear the the examples in your link are not compatible with Mesa 2.1.5. Running python and then giving the command:
>>>import run
from the appropriate directory produces the error:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\mesa-examples\examples\forest_fire\run.py", line 1, in <module>
    from forest_fire.server import server
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\mesa-examples\examples\forest_fire\forest_fire\server.py", line 3, in <module>
    from .model import ForestFire
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\mesa-examples\examples\forest_fire\forest_fire\model.py", line 3, in <module>
    from .agent import TreeCell
  File "C:\Users\Nick\anaconda3\envs\mesaenv39\mesa-examples\examples\forest_fire\forest_fire\agent.py", line 4, in <module>
    class TreeCell(mesa.Agent):
AttributeError: module 'mesa' has no attribute 'Agent'

I can get rid of that particular error by editing agent.py, but then another arises... and so on. It seems clear the directory structure has been changed, but without knowing how, or what Mesa version the examples belong to, I'm stuck. I want to be able to run examples in a recent version in order to have another try at updating my model which runs in version 0.8.8.1, but I can't do this without a working example under a more recent version. Can anyone assist? Thanks.

**Reply by EwoutH** [reply]
I wasn't involved with moving the examples to [mesa-examples](https://github.com/projectmesa/mesa-examples), but maybe can @tpike3 @jackiekazil or @rht assist.

I suspect it has something to do with from what directory you try to run them (so something with import paths).

**Reply by NickGotts** [reply]
Well possibly - and thanks for the prompt response. But I've tried placing mesa-examples in various relationships to the mesa directories, and there's nothing in sys.path that suggests to me that the system is looking in the wrong place.

**Comment by jackiekazil** [comment]
Looks like the requirements files for Forest Fire wasn't updated. I updated it in this PR -- https://github.com/projectmesa/mesa-examples/pull/75

I found out there a few more like this -- https://github.com/projectmesa/mesa-examples/issues/76

One the PR is reviewed and merged that solves FF... I will try to do the rest this evening. 
Please let us know if you need anything else.

**Reply by jackiekazil** [reply]
@NickGotts if you want to make it work manually -- install mesa 2+... the code works with 2+ but the requirements file wasn't updated in the update. That was an error.

**Reply by EwoutH** [reply]
Thanks for figuring this out!

Let's add this to the agenda for Thursday:
- https://github.com/projectmesa/mesa/discussions/1929

**Comment by NickGotts** [comment]
Thanks very much, but I'm not entirely clear what I need to do. I've got mesa 2.1.5 installed, and the example I've mostly been trying (because it uses continuous space) is boid_flockers, which you don't mention as needing any updating. I'll try setting up a new environment, reinstalling 2.1.5, then "git clone https://github.com/projectmesa/mesa-examples/" and "mesa runserver" in examples\boid_flockers. As far as you know, should that work? (Incidentally, I have no experience with Colab, but I don't think it would be useful for what I want to do, which is update my model CINCH to a recent mesa version. CINCH has around 2,500 lines of code, has fairly complicated input files, and runs in three different ways - single runs with a visualization, using mesa's batch-running facilities, and using the EMA workbench. I did try updating to 2.1.1 before, but couldn't get the visualization to work.)

**Comment by NickGotts** [comment]
OK - this time boid_flockers worked (and so did forest_fire). The only thing I did differently was where I gave the clone command. So thanks for your assistance, particularly given the time of year! My plan now is to see if the version of boid_flockers that runs under mesa 0.8.8.1 runs under 2.1.5, and if not, what has to be changed to make it do so; and if I can do that, try making parallel changes to CINCH and see if that will run under 2.1.5. So I may have further issues, given that my previous attempt failed.

**Comment by NickGotts** [comment]
sigh...
Unbelievably, the visualization for boid_flockers is now failing to work once more. That for forest_fire still works. I've tried setting up a new Anaconda3 Python 3.9 environment, moving to the directory containing that version of Python, then:

pip install mesa
git clone https://github.com/projectmesa/mesa-examples/

- both apparently successful.

cd mesa-examples\examples\forest_fire
mesa runserver

- This runs correctly.

cd ..\boid_flockers
mesa runserver

- A "visualization" is opened, but only shows the top lines, down to and including:
 Current Step 0
- The canvas does not appear. If "start is then clicked, the model apparently runs, the Current Step increases (and the steps are also shown on the anaconda environment terminal), but still, no canvas. Halting the model and returning to forest_fire, that again works correctly. Back to boid-flockers: same as before, no canvas. The examples color_patches, conways_game_of_life and wolf_sheep all run correctly.

Does anyone have any idea what is going on here? Can anyone reproduce this behaviour? Seems like it is something specific to boid_flockers, or more generally to visualization of continuous space.

**Reply by jackiekazil** [reply]
I followed your steps. I also did this in an isolated environment. I wasn't able to recreate the issue. 

This shouldn't impact the examples you are working in, but we are currently working in an experimental front-end -- samples of this can found in examples ending in with "_experimental". I don't believe any continuous space ones have been created yet though. (/cc @tpike3)

**Comment by NickGotts** [comment]
Bizarre... shutting my desktop down then rebooting itrestored the visualization. I'll continue to experiment.

**Comment by NickGotts** [comment]
I suspect the problem arises from some interaction between the browser (I'm using Firefox) and the two versions of mesa (0.8.8.1 and 2.1.5) I'm working with - and possibly Anaconda as well. But I have not found a way of reproducing the problem consistently - sometimes it occurs, sometimes it doesn't. Last time it occurred, shutting down the browser and restarting it resolved the problem, but I don't know if this will always work. I'll report it here if I find a way of reproducing the problem consistently.

**Comment by NickGotts** [comment]
Finally got the CINCH visualization working in 2.1.5! The problem was with the simple_continuous_canvas.ps file. The 0.8.8.1 version has "var" in several places where the 2.1.5 version has "const". I don't know why (or in what version) this change was made, nor why using the 0.8.8.1 version under 2.1.5 doesn't work (because I had added code to the 0.8.8.1 version to get the visualization to show walls, doors and beds - the CINCH environment is a hospital ward - I copied that version across. I think the occasions when boid_flockers failed to work for me in 2.1.5 must have been due to it somehow accessing the 0.8.8.1 version - I admit my undertanding of the roles and operation of sys.path, __pycache__ and local_includes (I suspect one or more of these was involved), as well as of javascript, is rudimentary.

**Reply by jackiekazil** [reply]
This sounds like it is resolved then, yes? ... for now... 🤞
